### PR TITLE
[SPARK-11829] [ML] Add read/write to estimators under ml.feature (II)

### PR DESCRIPTION
**Pull request by yanboliang**
Add read/write support to the following estimators under spark.ml:
- ChiSqSelector
- PCA
- VectorIndexer
- Word2Vec

### CODE REVIEW

**Comment by yanboliang** [comment]
Jenkins, test this please.

File: mllib/src/test/scala/org/apache/spark/ml/feature/PCASuite.scala [hunk 1]
```diff
@@ -65,4 +65,24 @@ class PCASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    val data = Seq(
+      (0.0, Vectors.sparse(5, Seq((1, 1.0), (3, 7.0)))),
+      (1.0, Vectors.dense(2.0, 0.0, 3.0, 4.0, 5.0)),
+      (2.0, Vectors.dense(4.0, 0.0, 0.0, 6.0, 7.0))
+    )
+    val df = sqlContext.createDataFrame(data).toDF("id", "features")
+    val pca = new PCA().setK(3)
+    testEstimatorAndModelReadWrite(pca, df, allParams, checkModelData)
+  }
 }
```
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/feature/PCASuite.scala]: The same as [StandardScalerModel](https://github.com/apache/spark/pull/9839/files#diff-7ae15fe7b1bc41e6a876c2424380d764R132), we can not construct PCAModel directly by specifying the variable ```k```, so test estimator and model in one case with ```testEstimatorAndModelReadWrite```.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 2]
```diff
@@ -432,9 +432,9 @@ class Word2Vec extends Serializable with Logging {
  *                    (i * vectorSize, i * vectorSize + vectorSize)
  */
 @Since("1.1.0")
-class Word2VecModel private[mllib] (
-    private val wordIndex: Map[String, Int],
-    private val wordVectors: Array[Float]) extends Serializable with Saveable {
+class Word2VecModel private[spark] (
+    val wordIndex: Map[String, Int],
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Shall we keep them `private[spark]` for this release? We should not introduce unplanned APIs at this stage.

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala [hunk 3]
```diff
@@ -20,14 +20,16 @@ package org.apache.spark.ml.feature
 import java.lang.{Double => JDouble, Integer => JInt}
 import java.util.{Map => JMap}
 
+import org.apache.hadoop.fs.Path
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala]: organize imports

**Comment by mengxr** [comment]
LGTM except two minor issues commented inline.

**Comment by mengxr** [comment]
Merged into master and branch-1.6. Thanks!

**Pull request closed without merging**